doc/BMIconventions.md

```diff
@@ -0,0 +1,313 @@
+
+This means among other things that it is not possible to "extend" the BMI interface simply by adding functions to the base interface or header files. In any case, even with subclassed BMI models in Python and C++ with additional functions, ngen would not call any such additional functions because it does not know that they exist. 
+
+# Time Control
+
+## Set and Get Values Per Timestep 
+
+Importantly, for your model to work in ngen, all input values and output values pertain to the processing to be done *in a single engine timestep*. Put another way, a model running in ngen should always follow this general execution pattern:
+
+ * `initialize(...)` and any other setup at the beginning of the model, such as `get_input_var_names(...)`
+ * Now we will loop over the whole simulation time one timestep per iteration...
+   * `set_value(...)` set the inputs for the next timestep
+   * `update_until(...)` increment to the next timestep--computation happens here
+   * `get_value(...)` get the model outputs for the just-processed timestep
+
+Other patterns may be allowed by the BMI without being prescriptive, such as passing a whole time series of inputs as an array to `set_value(...)` but this is not supported by ngen. Passing of arrays as input is supported but is applicable in cases where the whole array applies to one timestep (e.g. soil moisture profile or snow layer information for multiple depths).
+
+## `update` vs. `update_until`
+
+For the foreseeable future, ngen will always call `update_until(...)` and will not call a model's `update()` method. Per the [BMI documentation](https://bmi.readthedocs.io/en/stable/#update-until) for `update_until`, an absolute time is passed as the argument, not an incremental time; i.e., if the model start time is `0.0` and ngen is incrementing timesteps by 3600 seconds, the first three iterations will call `update_until` with the values `3600.0`, `7200.0`, and `10800.0`.
+
+At present, ngen will always pass a value greater than the previous value passed to `update_until` to any model instance (it will move only forward in time).
+
+## Time Units
+
+At present, all BMI modules must use seconds as the time units. The output of `get_time_units()` should be `seconds` and the value passed to `update_until(...)` should be interpreted as seconds past time `0` (see below). This is something that we intend to make more flexible in the future.
+
```

Comment:
```suggestion
```
Obviated by #633

Reply:
Noting that follow-on discussion noted that BMI documentation is very vague about what should be passed to `update_until`, and whether the value returned by `get_time_units()` applies to `update_until()` ... recommend leaving this as is.